webui/react/src/components/kit/LineChart.tsx

```diff
@@ -310,6 +315,71 @@ const VirtualChartRenderer: React.FC<
   );
 };
 
+export const calculateChartProps = (
+  metricData: TrialMetricData,
+  experiments: ExperimentWithTrial[],
+  trials: TrialItem[],
+  xAxis: XAxisDomain,
+  colorMap: MapOfIdsToColors,
+): Loadable<ChartProps> => {
+  const { metricHasData, metrics, data, isLoaded, selectedMetrics } = metricData;
+  const chartedMetrics: Record<string, boolean> = {};
+  const out: ChartsProps = [];
+  const expNameById: Record<number, string> = {};
+  experiments.forEach((e) => {
+    expNameById[e.experiment.id] = e.experiment.name;
+  });
+  metrics.forEach((metric) => {
+    const series: Serie[] = [];
+    const key = `${metric.type}|${metric.name}`;
+    trials.forEach((t) => {
+      const m = data[t?.id || 0];
+      m?.[key] &&
+        t &&
+        series.push({
+          ...m[key],
+          color: colorMap[t.experimentId],
+          metricType: undefined,
+          name: expNameById[t.experimentId]
+            ? `${expNameById[t.experimentId]} (${t.experimentId})`
+            : String(t.experimentId),
+        });
+      chartedMetrics[key] ||= series.length > 0;
+    });
+    out.push({
+      series: Loaded(series),
+      title: <MetricBadgeTag metric={metric} />,
+      xAxis,
+      xLabel: String(xAxis),
+    });
+  });
+
+  // In order to show the spinner for each chart in the ChartGrid until
+  // metrics are visible, we must determine whether the metrics have been
+  // loaded and whether the chart props have been updated.
+  // If any metric has data but no chartProps contain data for the metric,
+  // then the charts have not been updated and we need to continue to show the
+  // spinner.
+  const chartDataIsLoaded = metrics.every((metric) => {
+    const metricKey = `${metric.type}|${metric.name}`;
+    return metricHasData?.[metricKey] ? !!chartedMetrics?.[metricKey] : true;
+  });
+  if (!isLoaded) {
+    // When trial metrics hasn't loaded metric names or individual trial metrics.
+    return NotLoaded;
+  } else if (!chartDataIsLoaded || !isEqual(selectedMetrics, metrics)) {
```

Comment:
the reason I added the `!isEqual(selectedMetrics, metrics))` check is because one bug I was seeing is that when selecting new experiments sometimes for a split second the UI would show `no data to plot`. I tracked down the issue being that:

1. A user selects a new experiment 
2. The new metric names are queried for both the current experiments and new one
3. Eventually a call to `summarizedMetricToSeries` is made but the input metrics do not include all the new metrics from (1). 
4. We show `no data to plot` since we think the new metrics from (1) do not have any data, but actually we have not searched for them yet. 

The `isEqual` check ensures that the metrics that were searched for matches the currently selected metrics.